### ISSUE TITLE
Transformer encoder with rnn attention ecoder is not supported yet

### ISSUE THREAD

**Issue by wxt1997** [espnet/espnet#2166]
May I ask the reason for ESPnet does not supported Transformer encoder in rnnt-att is the attention type will include both `MultiHeadedAttention` and attention types in `espnet/nets/pytorch_backend/rnn/attentions` ?  

And an error occured when I run a `etype=Transformer`, `dtype=Transformer`, `rnnt_mode=rnnt` transducer model and returned:   
```
File "/home/oshindo/espnet/espnet/asr/asr_utils.py", line 231, in get_attention_weight  
    if len(att_w.shape) == 3:  
AttributeError: 'str' object hs no attribute 'shape'  
```
I don't understand why the `att_w` is a 'str' type.

**Comment by wxt1997** [comment]
It seems that the return from `e2e_asr_transducer.calculate_all_attentions` is a `dict`, and when `att_w` is emumberated from `att_ws`, it becomes a 'str' which is a layer name.

**Comment by wxt1997** [comment]
I already fixed it by pointing `att_ws` is a `dict` in `/home/oshindo/espnet/espnet/asr/asr_utils.py`, maybe it's a bug.

**Comment by b-flo** [comment]
> May I ask the reason for ESPnet does not supported Transformer encoder in rnnt-att is the attention type will include both MultiHeadedAttention and attention types in espnet/nets/pytorch_backend/rnn/attentions ?

I don't recall the specific reason but it was due to some incompatibility between `MultiheadedAttention` and other attention used in RNN decoder. I 

> And an error occured when I run a etype=Transformer, dtype=Transformer, rnnt_mode=rnnt transducer model and returned:

```
File "/home/oshindo/espnet/espnet/asr/asr_utils.py", line 231, in get_attention_weight  
    if len(att_w.shape) == 3:  
AttributeError: 'str' object hs no attribute 'shape'  
```

In master branch? That is weird, the only time I met this error was... well yesterday evening in an integration test when I push a fix related to our discussion in #2162. But it's not merged.
The problem start within `asr.py`, I'm currently working on it.

**Comment by wxt1997** [comment]
Really thanks for your awsome work!
 > I don't recall the specific reason but it was due to some incompatibility between `MultiheadedAttention` and other attention used in RNN decoder usage. I may enable it in the future (didn't have time to work on it and it's not a priority)  

What is the meaning of incompatibility? I thought it will be easier to find an attention-type like `AttAdd` in `rnn/attentions` with `MultiheadedAttention` together in `e2e_asr_transducer.calculate_all_attentions`.

> In master branch? That is weird, the only time I met this error was... well yesterday evening in an integration test when I pushed a fix related to our discussion in #2162. But it's not merged and the problem is on Transformer-Transducer-Att because a method from the wrong class is called.
> The problem starts within `asr.py`, I'm currently working on it.

Emm, actually it's not a big deal. I just add a sentence like `att_w_npy = att_w[key]` in `asr/asr_utils.PlotAttentionReport.get_attention_weights`. Then the project works again. Do you mean that the problem starts with `asr.py` is `att_vis_fn = model.calculate_all_attentions`?

**Comment by b-flo** [comment]
Sorry, I had other things to deal with.

> What is the meaning of incompatibility? I thought it will be easier to find an attention-type like AttAdd in rnn/attentions with MultiheadedAttention together in e2e_asr_transducer.calculate_all_attentions.

What I mean by "incompatibility" is that some issues occurred due to the use of MultiheadedAttention coupled to other attentions, but I can't recall. I may investigate after my current PR, to be honest it wasn't a priority at that time.

> Sorry, I don't really know how to get a fixed project. The branch I used is download in a couple days ago by git clone. Emm, actually it's not a big deal. I just add a sentence like att_w_npy = att_w[key] in asr/asr_utils.PlotAttentionReport.get_attention_weights. Then the project works again. Do you mean that the problem starts with asr.py is att_vis_fn = model.calculate_all_attentions?

In `asr.py`, these conditions seems wrong to me: https://github.com/espnet/espnet/blob/master/espnet/asr/pytorch_backend/asr.py#L662-L664

It disable attention weights saving for all cases because `mtl_mode` was not set to "transducer" as you mentionned in #2162. However modifying the condition shows another problem, when I unified R, I discarded a decorator used only for transformer: https://github.com/espnet/espnet/blob/master/espnet/nets/pytorch_backend/e2e_asr_transformer.py#L204-L207.
It's fixed now but it's kind of an hack though.
